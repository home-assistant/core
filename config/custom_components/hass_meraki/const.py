"""Constants for the Meraki integration."""

DOMAIN = "hass_meraki"
