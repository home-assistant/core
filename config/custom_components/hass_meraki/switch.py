"""The switch Component."""

import logging

import meraki

from homeassistant.components.switch import SwitchEntity
import homeassistant.helpers.device_registry as dr
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from . import DOMAIN

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up Meraki sensors via config entry."""
    coordinator = hass.data[DOMAIN]["coordinator"]
    device_registry = dr.async_get(hass)
    config_entry = hass.data[DOMAIN].get("config_entry")
    if not config_entry:
        _LOGGER.error("Config entry not found!")
        return

    entities = []
    # Erstelle Sensoren für Geräte, die Client-Daten unterstützen (z. B. Switch, Access Point, Appliance)
    for serial, device_data in coordinator.data.items():
        # Registriere das Gerät im Device Registry
        device_registry.async_get_or_create(
            config_entry_id=config_entry.entry_id,
            identifiers={(DOMAIN, serial)},
            manufacturer="Cisco Meraki",
            name=device_data.get("name", f"Meraki Device {serial}"),
            model=device_data.get("model", "Unknown"),
            sw_version=device_data.get("firmware", "Unknown"),
            connections={("mac", device_data.get("mac", "Unknown"))},
        )

        for port in device_data.get("ports"):
            if port["poe"]["isAllocated"]:
                entities.append(MerakiPoeSwitch(coordinator, serial, port))  # noqa: PERF401
    async_add_entities(entities)

    # Optionally, subscribe to updates to add or remove entities dynamically.
    # Dynamic entity management can be more involved—this is a simplified approach.


class MerakiPoeSwitch(CoordinatorEntity, SwitchEntity):  # noqa: D101
    def __init__(self, coordinator, serial, port):
        """Initialize the switch entity for a specific port."""
        device_data = coordinator.data.get(serial, {})
        super().__init__(coordinator)
        self._serial = serial
        if "lldp" in port:
            self._attr_name = f"Port {port['portId']} {port['lldp']['systemName']}"
        else:
            self._attr_name = f"Port {port['portId']}"
        self._attr_unique_id = f"{serial}_poe_port_{port['portId']}"
        self._attr_has_entity_name = True
        self._port_id = port["portId"]
        self._enabled = port["enabled"]

        # Gerätedaten, damit die Entität dem Gerät zugeordnet wird
        self._attr_device_info = {
            "identifiers": {(DOMAIN, serial)},
            "name": device_data.get("name", f"Meraki Device {serial}"),
            "manufacturer": "Cisco Meraki",
            "model": device_data.get("model", "Unknown"),
            "sw_version": device_data.get("firmware", "Unknown"),
            "connections": {("mac", device_data.get("mac", "Unknown"))},
        }

    @property
    def is_on(self):
        """Return True if this port is enabled."""
        # Assume the API returns a boolean 'enabled' flag
        return self._enabled

    async def async_turn_on(self, **kwargs):
        """Enable this port."""
        result = await self.hass.async_add_executor_job(self._set_poe, True)
        if result:
            self._enabled = result.get("enabled")
            self.async_write_ha_state()

    async def async_turn_off(self, **kwargs):
        """Disable this port."""
        result = await self.hass.async_add_executor_job(self._set_poe, False)
        if result:
            self._enabled = result.get("enabled")
            self.async_write_ha_state()

    def _set_poe(self, enable):
        """Call the Meraki API to change the state for this port."""
        # Adjust the API call to match your needs. This is an example.

        api_key = self.coordinator.config_entry.data["api_key"]
        dashboard = meraki.DashboardAPI(api_key=api_key, suppress_logging=True)
        try:
            # This is a placeholder; replace with the actual API method and parameters.
            return dashboard.switch.updateDeviceSwitchPort(
                self._serial, self._port_id, enabled=enable
            )
        except meraki.APIError as err:
            _LOGGER.error("Failed to set state on port %s: %s", self._port_id, err)
            return False

    # @property
    # def extra_state_attributes(self):
    #     """Optional: expose additional port info as attributes."""
    #     return {"port_name": self._port.get("name")}
