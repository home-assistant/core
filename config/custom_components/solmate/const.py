"""Constants for the solmate integration."""

DOMAIN = "solmate"
