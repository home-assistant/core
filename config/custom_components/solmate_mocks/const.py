"""Constants for the solmate_mocks integration."""

DOMAIN = "solmate_mocks"
