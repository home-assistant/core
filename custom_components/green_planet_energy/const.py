"""Constants for the Green Planet Energy integration."""

DOMAIN = "green_planet_energy"

# Config flow keys
CONF_API_URL = "api_url"

# Default values
DEFAULT_API_URL = "https://mein.green-planet-energy.de/p2"
DEFAULT_SCAN_INTERVAL = 60  # minutes

# API constants
API_METHOD = "getVerbrauchspreisUndWindsignal"
AGGREGATS_ZEITRAUM = ""
AGGREGATS_TYP = ""
SOURCE = "Portal"

# Sensor keys for each hour (9-18)
SENSOR_HOURS = [9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
