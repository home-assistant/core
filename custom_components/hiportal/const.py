"""Constants for hiportal."""

DOMAIN = "hiportal"

CONF_USERNAME = "username"
CONF_PASSWORD = "password"
