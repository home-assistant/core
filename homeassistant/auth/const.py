"""Constants for the auth module."""

from datetime import timedelta

ACCESS_TOKEN_EXPIRATION = timedelta(minutes=30)
MFA_SESSION_EXPIRATION = timedelta(minutes=5)
REFRESH_TOKEN_EXPIRATION = timedelta(days=90).total_seconds()

GROUP_ID_ADMIN = "system-admin"
GROUP_ID_USER = "system-users"
GROUP_ID_READ_ONLY = "system-read-only"
