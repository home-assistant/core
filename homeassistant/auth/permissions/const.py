"""Permission constants."""

CAT_ENTITIES = "entities"
CAT_CONFIG_ENTRIES = "config_entries"
SUBCAT_ALL = "all"

POLICY_READ = "read"
POLICY_CONTROL = "control"
POLICY_EDIT = "edit"
