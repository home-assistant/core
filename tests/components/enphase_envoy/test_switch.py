"""Test Enphase Envoy switch platform."""

from unittest.mock import AsyncMock, patch

import pytest
from syrupy.assertion import SnapshotAssertion

from homeassistant.components.enphase_envoy.const import Platform
from homeassistant.components.switch import DOMAIN as SWITCH_DOMAIN
from homeassistant.const import (
    ATTR_ENTITY_ID,
    SERVICE_TOGGLE,
    SERVICE_TURN_OFF,
    SERVICE_TURN_ON,
    STATE_OFF,
    STATE_ON,
)
from homeassistant.core import HomeAssistant
from homeassistant.helpers import entity_registry as er

from . import setup_integration

from tests.common import MockConfigEntry, snapshot_platform


@pytest.mark.parametrize(
    ("mock_envoy"),
    ["envoy_metered_batt_relay", "envoy_eu_batt"],
    indirect=["mock_envoy"],
)
@pytest.mark.usefixtures("entity_registry_enabled_by_default")
async def test_switch(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    mock_envoy: AsyncMock,
    config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test switch platform entities against snapshot."""
    with patch("homeassistant.components.enphase_envoy.PLATFORMS", [Platform.SWITCH]):
        await setup_integration(hass, config_entry)
    await snapshot_platform(hass, entity_registry, snapshot, config_entry.entry_id)


@pytest.mark.parametrize(
    ("mock_envoy"),
    [
        "envoy",
        "envoy_1p_metered",
        "envoy_nobatt_metered_3p",
        "envoy_tot_cons_metered",
    ],
    indirect=["mock_envoy"],
)
async def test_no_switch(
    hass: HomeAssistant,
    mock_envoy: AsyncMock,
    config_entry: MockConfigEntry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test switch platform entities are not created."""
    with patch("homeassistant.components.enphase_envoy.PLATFORMS", [Platform.SWITCH]):
        await setup_integration(hass, config_entry)
    assert not er.async_entries_for_config_entry(entity_registry, config_entry.entry_id)


@pytest.mark.parametrize(
    ("mock_envoy"), ["envoy_metered_batt_relay"], indirect=["mock_envoy"]
)
async def test_switch_grid_operation(
    hass: HomeAssistant,
    mock_envoy: AsyncMock,
    config_entry: MockConfigEntry,
) -> None:
    """Test switch platform operation for grid switches."""
    with patch("homeassistant.components.enphase_envoy.PLATFORMS", [Platform.SWITCH]):
        await setup_integration(hass, config_entry)

    sn = mock_envoy.data.enpower.serial_number
    test_entity = f"{Platform.SWITCH}.enpower_{sn}_grid_enabled"

    # validate envoy value is reflected in entity
    assert (entity_state := hass.states.get(test_entity))
    assert entity_state.state == STATE_ON

    # test grid status switch operation
    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_OFF,
        {ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    mock_envoy.go_off_grid.assert_awaited_once_with()
    mock_envoy.go_off_grid.reset_mock()

    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_ON,
        {ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    mock_envoy.go_on_grid.assert_awaited_once_with()
    mock_envoy.go_on_grid.reset_mock()

    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TOGGLE,
        {ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    mock_envoy.go_off_grid.assert_awaited_once_with()
    mock_envoy.go_off_grid.reset_mock()


@pytest.mark.parametrize(
    ("mock_envoy", "use_serial"),
    [
        ("envoy_metered_batt_relay", "enpower_654321"),
        ("envoy_eu_batt", "envoy_1234"),
    ],
    indirect=["mock_envoy"],
)
async def test_switch_charge_from_grid_operation(
    hass: HomeAssistant,
    mock_envoy: AsyncMock,
    config_entry: MockConfigEntry,
    use_serial: str,
) -> None:
    """Test switch platform operation for charge from grid switches."""
    with patch("homeassistant.components.enphase_envoy.PLATFORMS", [Platform.SWITCH]):
        await setup_integration(hass, config_entry)

    test_entity = f"{Platform.SWITCH}.{use_serial}_charge_from_grid"

    # validate envoy value is reflected in entity
    assert (entity_state := hass.states.get(test_entity))
    assert entity_state.state == STATE_ON

    # test grid status switch operation
    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_OFF,
        {ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    mock_envoy.disable_charge_from_grid.assert_awaited_once_with()
    mock_envoy.disable_charge_from_grid.reset_mock()

    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TURN_ON,
        {ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    mock_envoy.enable_charge_from_grid.assert_awaited_once_with()
    mock_envoy.enable_charge_from_grid.reset_mock()

    await hass.services.async_call(
        SWITCH_DOMAIN,
        SERVICE_TOGGLE,
        {ATTR_ENTITY_ID: test_entity},
        blocking=True,
    )
    mock_envoy.disable_charge_from_grid.assert_awaited_once_with()
    mock_envoy.disable_charge_from_grid.reset_mock()


@pytest.mark.parametrize(
    ("mock_envoy", "entity_states"),
    [
        (
            "envoy_metered_batt_relay",
            {
                "NC1": (STATE_OFF, 0, 1),
                "NC2": (STATE_ON, 1, 0),
                "NC3": (STATE_OFF, 0, 1),
            },
        )
    ],
    indirect=["mock_envoy"],
)
async def test_switch_relay_operation(
    hass: HomeAssistant,
    mock_envoy: AsyncMock,
    config_entry: MockConfigEntry,
    entity_states: dict[str, tuple[str, int, int]],
) -> None:
    """Test enphase_envoy switch relay entities operation."""
    with patch("homeassistant.components.enphase_envoy.PLATFORMS", [Platform.SWITCH]):
        await setup_integration(hass, config_entry)

    entity_base = f"{Platform.SWITCH}."

    for contact_id, dry_contact in mock_envoy.data.dry_contact_settings.items():
        name = dry_contact.load_name.lower().replace(" ", "_")
        test_entity = f"{entity_base}{name}"
        assert (entity_state := hass.states.get(test_entity))
        assert entity_state.state == entity_states[contact_id][0]
        open_count = entity_states[contact_id][1]
        close_count = entity_states[contact_id][2]

        await hass.services.async_call(
            SWITCH_DOMAIN,
            SERVICE_TURN_OFF,
            {ATTR_ENTITY_ID: test_entity},
            blocking=True,
        )

        mock_envoy.open_dry_contact.assert_awaited_once_with(contact_id)
        mock_envoy.close_dry_contact.assert_not_awaited()
        mock_envoy.open_dry_contact.reset_mock()

        await hass.services.async_call(
            SWITCH_DOMAIN,
            SERVICE_TURN_ON,
            {ATTR_ENTITY_ID: test_entity},
            blocking=True,
        )

        mock_envoy.close_dry_contact.assert_awaited_once_with(contact_id)
        mock_envoy.open_dry_contact.assert_not_awaited()
        mock_envoy.close_dry_contact.reset_mock()

        await hass.services.async_call(
            SWITCH_DOMAIN,
            SERVICE_TOGGLE,
            {ATTR_ENTITY_ID: test_entity},
            blocking=True,
        )

        assert mock_envoy.open_dry_contact.await_count == open_count
        assert mock_envoy.close_dry_contact.await_count == close_count
        mock_envoy.open_dry_contact.reset_mock()
        mock_envoy.close_dry_contact.reset_mock()
