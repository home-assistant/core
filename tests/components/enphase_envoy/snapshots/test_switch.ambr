# serializer version: 1
# name: test_switch[envoy_eu_batt][switch.envoy_1234_charge_from_grid-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.envoy_1234_charge_from_grid',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Charge from grid',
    'platform': 'enphase_envoy',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'charge_from_grid',
    'unique_id': '1234_charge_from_grid',
    'unit_of_measurement': None,
  })
# ---
# name: test_switch[envoy_eu_batt][switch.envoy_1234_charge_from_grid-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Envoy 1234 Charge from grid',
    }),
    'context': <ANY>,
    'entity_id': 'switch.envoy_1234_charge_from_grid',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.enpower_654321_charge_from_grid-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.enpower_654321_charge_from_grid',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Charge from grid',
    'platform': 'enphase_envoy',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'charge_from_grid',
    'unique_id': '654321_charge_from_grid',
    'unit_of_measurement': None,
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.enpower_654321_charge_from_grid-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Enpower 654321 Charge from grid',
    }),
    'context': <ANY>,
    'entity_id': 'switch.enpower_654321_charge_from_grid',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.enpower_654321_grid_enabled-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.enpower_654321_grid_enabled',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Grid enabled',
    'platform': 'enphase_envoy',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'grid_enabled',
    'unique_id': '654321_mains_admin_state',
    'unit_of_measurement': None,
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.enpower_654321_grid_enabled-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Enpower 654321 Grid enabled',
    }),
    'context': <ANY>,
    'entity_id': 'switch.enpower_654321_grid_enabled',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.nc1_fixture-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.nc1_fixture',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'enphase_envoy',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': '654321_relay_NC1_relay_status',
    'unit_of_measurement': None,
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.nc1_fixture-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'NC1 Fixture',
    }),
    'context': <ANY>,
    'entity_id': 'switch.nc1_fixture',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.nc2_fixture-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.nc2_fixture',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'enphase_envoy',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': '654321_relay_NC2_relay_status',
    'unit_of_measurement': None,
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.nc2_fixture-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'NC2 Fixture',
    }),
    'context': <ANY>,
    'entity_id': 'switch.nc2_fixture',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.nc3_fixture-entry]
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'switch',
    'entity_category': None,
    'entity_id': 'switch.nc3_fixture',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'labels': set({
    }),
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': None,
    'platform': 'enphase_envoy',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': None,
    'unique_id': '654321_relay_NC3_relay_status',
    'unit_of_measurement': None,
  })
# ---
# name: test_switch[envoy_metered_batt_relay][switch.nc3_fixture-state]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'NC3 Fixture',
    }),
    'context': <ANY>,
    'entity_id': 'switch.nc3_fixture',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'off',
  })
# ---
